sdr_soapy.c

```diff
@@ -0,0 +1,374 @@
+        for (size_t i = 0; i < result.size; ++i) {
+            printf("%s=%s", result.keys[i], result.vals[i]);
+            if (i+1 < result.size) printf(", ");
+        }
+        printf("\n");
+    }
+    SoapySDRKwargs_clear(&result);
+
+    char* hw_key = SoapySDRDevice_getHardwareKey(SOAPY.dev);
+    if (hw_key) {
+        printf("soapy: hardware key is %s\n", hw_key);
+        SOAPY.hw_key = strdup(hw_key);
+        SoapySDR_free(hw_key);
+    }
+
+    if (SOAPY.channel) {
+        size_t supported_channels = SoapySDRDevice_getNumChannels(SOAPY.dev, SOAPY_SDR_RX);
+        if (SOAPY.channel >= supported_channels) {
+            fprintf(stderr, "soapy: device only supports %ld channels\n", supported_channels);
+            goto error;
+        }
+    }
+
+    length = 0;
+    char** available_antennas = SoapySDRDevice_listAntennas(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, &length);
+    fprintf(stderr, "soapy: available antennas: ");
+    for (size_t i = 0; i < length; i++) {
+        fprintf(stderr, "%s", available_antennas[i]);
+        if (i+1 < length) printf(", ");
+    }
+    fprintf(stderr, "\n");
+    SoapySDRStrings_clear(&available_antennas, length);
+
+    if (SoapySDRDevice_setSampleRate(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, Modes.sample_rate) != 0) {
+        fprintf(stderr, "soapy: setSampleRate failed: %s\n", SoapySDRDevice_lastError());
+        goto error;
+    }
+
+    if (SoapySDRDevice_setFrequency(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, Modes.freq, NULL) != 0) {
+        fprintf(stderr, "soapy: setFrequency failed: %s\n", SoapySDRDevice_lastError());
+        goto error;
+    }
+
+    bool has_agc = SoapySDRDevice_hasGainMode(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel);
+    if (SOAPY.enable_agc) {
+        if (has_agc) {
+            if (SoapySDRDevice_setGainMode(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, 1) != 0) {
+                fprintf(stderr, "soapy: setGainMode failed: %s\n", SoapySDRDevice_lastError());
+                goto error;
+            }
+        }
+        else {
+            fprintf(stderr, "soapy: device does not support enabling AGC\n");
+            goto error;
+        }
+    }
+    else if(has_agc) {
+        // by default, disable AGC
+        if (SoapySDRDevice_setGainMode(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, 0) != 0) {
+            fprintf(stderr, "soapy: setGainMode failed: %s\n", SoapySDRDevice_lastError());
+            goto error;
+        }
+    }
+
+    if (Modes.gain != MODES_DEFAULT_GAIN) {
+        double gain;
+        if (SOAPY.dev_sdrplay) {
+            gain = Modes.gain;
+        } else {
+            gain = Modes.gain / 10.0;
+        }
+        if (SoapySDRDevice_setGain(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, gain) != 0) {
+            fprintf(stderr, "soapy: setGain failed: %s\n", SoapySDRDevice_lastError());
+            goto error;
+        }
+    }
+
+    if (SoapySDRDevice_setBandwidth(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, (SOAPY.dev_sdrplay ? 5.0e6 : 3.0e6)) != 0) {
```

Comment:
I'd prefer to make bandwidth configurable as a command-line option, and then set _defaults_ based on hardware.

(cf https://github.com/flightaware/dump978/blob/master/soapy_source.cc#L239 as a similar thing in dump978's soapysdr code)

Reply:
I've added the --bandwidth option.